designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,336 @@
+
+### Output formats
+
+JSON notation will be used for hierarchical and structured data.
+Tabular data will be either in
+
+- CSV format (human-readable)
+
+- Apache Arrow format, which is a self-describing binary data format.
+An Apache Arrow parquet file consists of an schema followed by one or more rows of data.
+The schema describes the structure of the data objects in each row, allowing the downstream
+process to reconstitute structured objects properly.
+The Arrow libraries for R and Python will allow downstream analysis packages
+to process these files, either during inference or afterwards.
+
+- Unformatted - the binary values returned by methods on the algorithm and model;
+all responsibility for handling these outputs lies with the caller.
+
+The Stan CSV output format will remain available via CmdStan.
+
+### Input formats
+
+Outputs from one inference can become inputs to a subsequent one.
+The input formats used by the inference algorithms are:
+
+- data variable definitions in JSON or [rdump](https://mc-stan.org/docs/cmdstan-guide/rdump.html)
+- initial parameter variables in JSON or rdump
+- for the standalone generated quantities, inputs include the Stan CSV file which contains the sample from the fitted model.
+
+Information from one inference run which would become either data or parameter inits would require converting CSV files to JSON.
+For standalone generated quantities, a sample in Apache format would need to be converted to CSV format.
+We will use existing converter libraries to handle this.
+
+## Technical Specification
+
+To implement the features outlined in the Functional Specification
+will require changes in the `stan::services` and `stan::callbacks` layers.
+
+We will define two [C++ Enumeration](https://en.cppreference.com/w/cpp/language/enum)
+classes: `InfoType` and `OutputFormat`.
+
+- The values of enum class `InfoType` correspond to the information items listed in
+the functional specification, e.g., `LP` (log_prob), `sample`, `metric`, etc.
+This provides an extensible mechanism for adding future outputs.
+
+- The values of enum class `OutputFormat` correspond to available output formats.
+We expect to implement at least the following:
+
+    + `Arrow` - tabular data in Apache Arrow, other structured data is JSON
+    + `Csv` - tabular data in CSV, other structured data is JSON
+    + `Raw` - tabular data is output as vector of binary values
+    + `StanCsv` - legacy Stan CSV format
+
+We will define an `OutputWriter` class which and we will add new methods to
+the services layer calling functions which take a single argument `output_writer`u
+(instead of arguments `sample_writer` and `diagnostic_writer`).
+The `OutputWriter` class will be sub-classed by output format:
+`ArrowOutputWriter` `CsvOutputWriter`, `RawOutputWriter`, `StanCsvOutputWriter`.
```

Comment:
[comment, implementation detail]

I think you could even make a `TabularAndJsonOutputWriter` abstract class and have `ArrowOutputWriter` and the (non-legacy) `CSVOutputWriter` both extend from that, since they'll be re-using json for non-tabular. This is too fined grained for this document, just wanted to mention that there is the opportunity for that sort of thing